README.md
[![Build Status](https://travis-ci.org/n3okill/enfsmkdirp.svg)](https://travis-ci.org/n3okill/enfsmkdirp)
[![Build status](https://ci.appveyor.com/api/projects/status/9dk56b6mtkdlmm7q?svg=true)](https://ci.appveyor.com/project/n3okill/enfsmkdirp)
[![Codacy Badge](https://api.codacy.com/project/badge/grade/69090ad47de140bbb12d82a2f5a8bed3)](https://www.codacy.com/app/n3okill/enfsmkdirp)
[![Donate](https://www.paypalobjects.com/en_US/i/btn/btn_donate_SM.gif)](https://www.paypal.com/cgi-bin/webscr?cmd=_s-xclick&hosted_button_id=64PYTCDH5UNZ6)

[![NPM](https://nodei.co/npm/enfsmkdirp.png)](https://nodei.co/npm/enfsmkdirp/)

enfsmkdirp
=========
Module that add list functionality to node fs module

**enfs** stands for [E]asy [N]ode [fs]

This module is intended to work as a sub-module of [enfs](https://www.npmjs.com/package/enfs)


Description
-----------
This module will add a method that allows the creation o directories
and sub-directories with one command line. Add mkdir -p functionality to node fs module

- This module will add following methods to node fs module:
  * mkdirp
  * mkdirpSync
  
Usage
-----
`enfsmkdirp`

```js
    var enfsmkdirp = require("enfsmkdirp");
```

Errors
------
All the methods follows the node culture.
- Async: Every async method returns an Error in the first callback parameter
- Sync: Every sync method throws an Error.


Additional Methods
------------------
- [mkdirp](#mkdirp)
- [mkdirpSync](#mkdirpsync)


### mkdirp
  - **mkdirp(path, [options], callback)**

> Asynchronously create multiple directories levels

`path`
The path can be on the form of a string, an array or "brace-string"

[options]:
  * fs (Object): an alternative fs module to use (default will be [enfspatch](https://www.npmjs.com/package/enfspatch))
  * mode (String or Number): the mode that will be attributed to the directory being created


  - Path (String)

```js
    enfsmkdirp.mkdirp("/home/path/to/folder", function(err){
        //do something
    });
```

  - Path (Array)

```js
    enfsmkdirp.mkdirp(["/home/path/to/folder","/var/home/test"], function(err){
        //do something
    });
```

  - Path ("brace-string")

```js
    enfsmkdirp.mkdirp("./project/{development,production}/{public,css,private,test}", function(err){
        //do something
    });
```

### mkdirpSync
  - **mkdirpSync(path, [options])**

> Synchronously create multiple directories levels

`path`
The path can be on the form of a string, an array or "brace-string"

[options]:
  * fs (Object): an alternative fs module to use (default will be [enfspatch](https://www.npmjs.com/package/enfspatch))
  * mode (String or Number): the mode that will be attributed to the directory being created


  - Path (String)

```js
    enfsmkdirp.mkdirpSync("/home/path/to/folder");
```

  - Path (Array)

```js
    enfsmkdirp.mkdirpSync(["/home/path/to/folder","/var/home/test"]);
```

  - Path ("brace-string")

```js
    enfsmkdirp.mkdirpSync("./project/{development,production}/{public,css,private,test}");
```


License
-------

Creative Commons Attribution 4.0 International License

Copyright (c) 2016 Joao Parreira <joaofrparreira@gmail.com> [GitHub](https://github.com/n3okill)

This work is licensed under the Creative Commons Attribution 4.0 International License. 
To view a copy of this license, visit [CC-BY-4.0](http://creativecommons.org/licenses/by/4.0/).




CHANGELOG.md
1.0.0 / 2017-02-21
==================
  * code improvement
  * dependencies upgrade
  * major version
  
0.1.0 / 2016-04-06
==================
  * es6 update
  * change to return all created paths
  * dependencies update


0.0.1 / 2016-02-20
==================

  * first release


test/mkdirpMock.js
var nodePath = require("path"),
    nodeOs = require("os"),
    rimraf = require("rimraf"),
    enFs = require("enfspatch"),
    mkdirp = require("../"),
    Mock = require("mock-fs"),
    cwd = process.cwd();


describe("esnofsmkdirp mockfs", function() {
    var _0777, _0755, tmpPath, isWindows;
    tmpPath = nodePath.join(nodeOs.tmpdir(), "enfsmkdirpsync");
    _0777 = parseInt("0777", 8);
    _0755 = parseInt("0755", 8);
    isWindows = /^win/.test(process.platform);
    before(function() {
        if (!enFs.existStatSync(tmpPath)) {
            enFs.mkdirSync(tmpPath);
        }
        process.chdir(tmpPath);
    });
    afterEach(function() {
        rimraf.sync(tmpPath + nodePath.sep + "*");
    });
    after(function() {
        process.chdir(cwd);
        rimraf.sync(tmpPath);
    });
    describe("> async", function() {
        it("should test options fs", function(done) {
            var tmpFs, file, x, y, z;
            tmpFs = Mock.fs({});

            x = Math.floor(Math.random() * Math.pow(16, 4)).toString(16);
            y = Math.floor(Math.random() * Math.pow(16, 4)).toString(16);
            z = Math.floor(Math.random() * Math.pow(16, 4)).toString(16);

            tmpFs.isTmpFs = true;
            enFs.hasOwnProperty("isTmpFs").should.be.equal(false);
            tmpFs.hasOwnProperty("isTmpFs").should.be.equal(true);

            file = nodePath.join(tmpPath, "beep", "boop", [x, y, z].join(nodePath.sep));

            mkdirp.mkdirp(file, {fs: tmpFs, mode: _0755}, function(err) {
                (err === null).should.be.equal(true);
                tmpFs.stat(file, function(errStats, stat) {
                    (errStats === null).should.be.equal(true);
                    stat.isDirectory().should.be.equal(true);
                    if (!isWindows) {
                        (stat.mode & _0777).should.be.equal(_0755);
                    }
                    done();
                });
            });
        });
    });
    describe("> sync", function() {
        it("should test options fs sync", function(done) {
            var tmpFs, file, x, y, z;
            tmpFs = Mock.fs({});

            x = Math.floor(Math.random() * Math.pow(16, 4)).toString(16);
            y = Math.floor(Math.random() * Math.pow(16, 4)).toString(16);
            z = Math.floor(Math.random() * Math.pow(16, 4)).toString(16);

            tmpFs.isTmpFs = true;
            enFs.hasOwnProperty("isTmpFs").should.be.equal(false);
            tmpFs.hasOwnProperty("isTmpFs").should.be.equal(true);

            file = nodePath.join(tmpPath, "beep", "boop", [x, y, z].join(nodePath.sep));

            mkdirp.mkdirpSync(file, {fs: tmpFs, mode: _0755});
            tmpFs.stat(file, function(err, stat) {
                (err === null).should.be.equal(true);
                stat.isDirectory().should.be.equal(true);
                if (!isWindows) {
                    (stat.mode & _0777).should.be.equal(_0755);
                }
                done();
            });
        });
    });
});


test/mkdirpSync.js
/**
 * Created by JParreir on 05-10-2015.
 */
/* global afterEach, beforeEach, describe, it, after, before, process */

"use strict";

const nodePath = require("path");
const nodeOs = require("os");
const rimraf = require("rimraf");
const enFs = require("enfspatch");
const mkdirp = require("../");
const mkdirpUtil = require("../lib/util");
const cwd = process.cwd();


describe("enfsmkdirp sync", function() {
    const invalidWindowsDrive = "AB:\\";
    const tmpPath = nodePath.join(nodeOs.tmpdir(), "enfsmkdirpsync");
    const _0777 = parseInt("0777", 8);
    const _0755 = parseInt("0755", 8);
    const isWindows = /^win/.test(process.platform);
    before(function() {
        if (!enFs.existAccessSync(tmpPath)) {
            enFs.mkdirSync(tmpPath);
        }
        process.chdir(tmpPath);
    });
    afterEach(function() {
        rimraf.sync(tmpPath + nodePath.sep + "*");
    });
    after(function() {
        process.chdir(cwd);
        rimraf.sync(tmpPath);
    });

    it("should test mkdirp sync", function(done) {
        const x = Math.floor(Math.random() * Math.pow(16, 4)).toString(16);
        const y = Math.floor(Math.random() * Math.pow(16, 4)).toString(16);
        const z = Math.floor(Math.random() * Math.pow(16, 4)).toString(16);
        const file = nodePath.join(tmpPath, [x, y, z].join(nodePath.sep));
        (function() {
            mkdirp.mkdirpSync(file, _0755);
            enFs.stat(file, function(errStat, stat) {
                (errStat == null).should.be.equal(true);
                stat.isDirectory().should.be.equal(true);
                if (!isWindows) {
                    (stat.mode & _0777).should.be.equal(_0755);
                }
                done();
            });
        }).should.not.throw();
    });
    it("should test permissions sync", function(done) {
        const file = nodePath.join(tmpPath, (Math.random() * (1 << 30)).toString(16) + ".json");
        (function() {
            mkdirp.mkdirpSync(file, _0755);
            enFs.stat(file, function(err, stat) {
                (err === null).should.be.equal(true);
                stat.isDirectory().should.be.equal(true);
                if (!isWindows) {
                    (stat.mode & _0777).should.be.equal(_0755);
                }
                done();
            });
        }).should.not.throw();
    });
    it("should test root permissions sync", function(done) {
        (function() {
            mkdirp.mkdirpSync(tmpPath, _0755);
            enFs.stat(tmpPath, function(err, stat) {
                (err === null).should.be.equal(true);
                stat.isDirectory().should.be.equal(true);
                done();
            });
        }).should.not.throw();
    });
    it("should test return value sync", function() {
        let made;
        const x = Math.floor(Math.random() * Math.pow(16, 4)).toString(16);
        const y = Math.floor(Math.random() * Math.pow(16, 4)).toString(16);
        const z = Math.floor(Math.random() * Math.pow(16, 4)).toString(16);

        const file = nodePath.join(tmpPath, [x, y, z].join(nodePath.sep));

        // should return the first dir created.
        // By this point, it would be profoundly surprising if /tmp didn't
        // already exist, since every other test makes things in there.
        // Note that this will throw on failure, which will fail the test.
        (function() {
            made = mkdirp.mkdirpSync(file);
            made.should.be.equal(nodePath.join(tmpPath, x, y, z));

            // making the same file again should have no effect.
            (function() {
                made = mkdirp.mkdirpSync(file);
                (made === file).should.be.equal(true);
            }).should.not.throw();
        }).should.not.throw();
    });
    it("should test sync", function(done) {
        const x = Math.floor(Math.random() * Math.pow(16, 4)).toString(16);
        const y = Math.floor(Math.random() * Math.pow(16, 4)).toString(16);
        const z = Math.floor(Math.random() * Math.pow(16, 4)).toString(16);

        const file = nodePath.join(tmpPath, [x, y, z].join(nodePath.sep));

        (function() {
            mkdirp.mkdirpSync(file, _0755);
            enFs.stat(file, function(err, stat) {
                (err === null).should.be.equal(true);
                stat.isDirectory().should.be.equal(true);
                if (!isWindows) {
                    (stat.mode & _0777).should.be.equal(_0755);
                }
                done();
            });
        }).should.not.throw();
    });
    it("should test implicit mode from umask sync", function(done) {
        const x = Math.floor(Math.random() * Math.pow(16, 4)).toString(16);
        const y = Math.floor(Math.random() * Math.pow(16, 4)).toString(16);
        const z = Math.floor(Math.random() * Math.pow(16, 4)).toString(16);

        const file = nodePath.join(tmpPath, [x, y, z].join(nodePath.sep));

        (function() {
            mkdirp.mkdirpSync(file);
            enFs.stat(file, function(err, stat) {
                (err === null).should.be.equal(true);
                stat.isDirectory().should.be.equal(true);
                if (!isWindows) {
                    (stat.mode & _0777).should.be.equal(_0777);
                }
                done();
            });
        }).should.not.throw();
    });
    it("should test null byte filename", function() {
        const nullChar = "\0";

        const x = Math.floor(Math.random() * Math.pow(16, 4)).toString(16);
        const y = Math.floor(Math.random() * Math.pow(16, 4)).toString(16);
        const z = Math.floor(Math.random() * Math.pow(16, 4)).toString(16);

        const file = nodePath.join(tmpPath, [x, y, z].join(nodePath.sep), nullChar);

        (function() {
            mkdirp.mkdirpSync(file);
        }).should.throw({message: /string without null bytes/});
    });
    it("should test curly braces sync", function(done) {
        const file = `${nodePath.resolve(tmpPath)}/{production,dev}/{css,img,js}`;
        const paths = mkdirpUtil.inspectCurlyBraces(file);
        let size = paths.length;

        (function() {
            mkdirp.mkdirpSync(file);
            paths.forEach(function(path) {
                enFs.stat(nodePath.resolve(path), function(errStat, stat) {
                    (errStat === null).should.be.equal(true);
                    stat.isDirectory().should.be.equal(true);
                    if (--size === 0) {
                        done();
                    }
                });
            });
        }).should.not.throw();
    });
    it("should test invalid path drive on windows sync", function() {
        if (!isWindows) {
            return;
        }
        const file = nodePath.join(invalidWindowsDrive, "fooSync");
        (function() {
            mkdirp.mkdirpSync(file)
        }).should.throw({code: "EINVALID", message: /Invalid character found in path./});
    });
    it("should test invalid filename with double quote sync", function() {
        const file = nodePath.join(tmpPath, `foo"bar`);
        if (!isWindows) {
            return;
        }
        (function() {
            mkdirp.mkdirpSync(file);
        }).should.throw({message: /Invalid character found/});
    });
});


test/mkdirpAsync.js
/**
 * Created by JParreir on 05-10-2015.
 */
/* global afterEach, beforeEach, describe, it, after, before, process */

"use strict";

const nodePath = require("path");
const nodeOs = require("os");
const rimraf = require("rimraf");
const enFs = require("enfspatch");
const mkdirp = require("../");
const mkdirpUtil = require("../lib/util");
const cwd = process.cwd();


describe("enfsmkdirp async", function() {
    const invalidWindowsDrive = "AB:\\";
    const tmpPath = nodePath.join(nodeOs.tmpdir(), "enfsmkdirpasync");
    const _0777 = parseInt("0777", 8);
    const _0755 = parseInt("0755", 8);
    const _0744 = parseInt("0744", 8);
    const isWindows = /^win/.test(process.platform);
    before(function() {
        if (!enFs.existAccessSync(tmpPath)) {
            enFs.mkdirSync(tmpPath);
        }
        process.chdir(tmpPath);
    });
    afterEach(function() {
        rimraf.sync(tmpPath + nodePath.sep + "*");
    });
    after(function() {
        process.chdir(cwd);
        rimraf.sync(tmpPath);
    });

    it("should test mkdirp", function(done) {
        const x = Math.floor(Math.random() * Math.pow(16, 4)).toString(16);
        const y = Math.floor(Math.random() * Math.pow(16, 4)).toString(16);
        const z = Math.floor(Math.random() * Math.pow(16, 4)).toString(16);
        const file = nodePath.join(tmpPath, [x, y, z].join(nodePath.sep));
        mkdirp.mkdirp(file, _0755, function(err) {
            (err === null).should.be.equal(true);
            enFs.stat(file, function(errStat, stat) {
                (errStat == null).should.be.equal(true);
                stat.isDirectory().should.be.equal(true);
                if (!isWindows) {
                    (stat.mode & _0777).should.be.equal(_0755);
                }
                done();
            });
        });
    });
    it("should test chmod", function(done) {
        const ps = [tmpPath];
        for (let i = 0; i < 25; i++) {
            ps.push(Math.floor(Math.random() * Math.pow(16, 4)).toString(16));
        }
        const file = ps.join(nodePath.sep);
        mkdirp.mkdirp(file, _0744, function(err) {
            (err === null).should.be.equal(true);
            enFs.stat(file, function(errStat, stat) {
                (errStat === null).should.be.equal(true);
                stat.isDirectory().should.be.equal(true);
                if (!isWindows) {
                    (stat.mode & _0777).should.be.equal(_0744);
                }
                mkdirp.mkdirp(file, _0755, function(err2) {
                    (err2 === null).should.be.equal(true);
                    enFs.stat(file, function(errStat2, stat2) {
                        (errStat2 === null).should.be.equal(true);
                        stat2.isDirectory().should.be.equal(true);
                        if (!isWindows) {
                            (stat2.mode & _0777).should.not.be.equal(_0755);
                            (stat2.mode & _0777).should.be.equal(_0744);
                        }
                        done();
                    });
                });
            });
        });
    });
    it("should test overwrite", function(done) {
        const ps = [tmpPath];
        for (let i = 0; i < 2; i++) {
            ps.push(Math.floor(Math.random() * Math.pow(16, 4)).toString(16));
        }
        const file = ps.join(nodePath.sep);
        //a file in the way
        const itw = ps.slice(0, 2).join(nodePath.sep);
        enFs.writeFile(itw, "I am in the way", function(errWrite) {
            (errWrite === null).should.be.equal(true);
            enFs.stat(itw, function(errStat, stat) {
                (errStat === null).should.be.equal(true);
                stat.isFile().should.be.equal(true);
                mkdirp.mkdirp(file, _0755, function(err) {
                    err.should.be.instanceOf(Error);
                    //using mkdir(2) will throw an EEXIST even if there's a file in the way instead of throwing ENOTDIR
                    err.code.should.be.equalOneOf("ENOTDIR", "EEXIST");
                    done();
                });
            });
        });
    });

    it("should test permissions", function(done) {
        const file = nodePath.join(tmpPath, (Math.random() * (1 << 30)).toString(16));

        mkdirp.mkdirp(file, _0755, function(err) {
            (err === null).should.be.equal(true);
            enFs.stat(file, function(errStat, stat) {
                (err === null).should.be.equal(true);
                stat.isDirectory().should.be.equal(true);
                if (!isWindows) {
                    (stat.mode & _0777).should.be.equal(_0755);
                }
                done();
            });
        });
    });
    it("should test root permissions", function(done) {
        mkdirp.mkdirp(tmpPath, _0755, function(err) {
            (err === null).should.be.equal(true);
            done();
        });
    });
    it("should test race", function(done) {
        const testsNumber = 10;
        let counter = testsNumber;
        const ps = [tmpPath];

        for (let i = 0; i < 25; i++) {
            ps.push(Math.floor(Math.random() * Math.pow(16, 4)).toString(16));
        }
        const file = ps.join(nodePath.sep);

        function makeFile(file) {
            mkdirp.mkdirp(file, _0755, function(err) {
                (err === null).should.be.equal(true);
                enFs.stat(file, function(errStat, stat) {
                    (errStat === null).should.be.equal(true);
                    stat.isDirectory().should.be.equal(true);
                    if (!isWindows) {
                        (stat.mode & _0777).should.be.equal(_0755);
                    }
                    if (--counter === 0) {
                        done();
                    }
                });
            });
        }

        for (let i = 0; i < testsNumber; i++) {
            makeFile(file);
        }
    });
    it("should test relative path", function(done) {
        const x = Math.floor(Math.random() * Math.pow(16, 4)).toString(16);
        const y = Math.floor(Math.random() * Math.pow(16, 4)).toString(16);
        const z = Math.floor(Math.random() * Math.pow(16, 4)).toString(16);

        const file = [x, y, z].join(nodePath.sep);

        mkdirp.mkdirp(file, _0755, function(err) {
            (err === null).should.be.equal(true);
            enFs.stat(file, function(errStat, stat) {
                (errStat === null).should.be.equal(true);
                stat.isDirectory().should.be.equal(true);
                if (!isWindows) {
                    (stat.mode & _0777).should.be.equal(_0755);
                }
                done();
            });
        });
    });
    it("should test relative path 2", function(done) {
        const x = Math.floor(Math.random() * Math.pow(16, 4)).toString(16);
        const y = Math.floor(Math.random() * Math.pow(16, 4)).toString(16);
        const z = Math.floor(Math.random() * Math.pow(16, 4)).toString(16);

        const file = nodePath.join("..","tmp",[x, y, z].join(nodePath.sep));

        mkdirp.mkdirp(file, _0755, function(err) {
            (err === null).should.be.equal(true);
            enFs.stat(file, function(errStat, stat) {
                (errStat === null).should.be.equal(true);
                stat.isDirectory().should.be.equal(true);
                if (!isWindows) {
                    (stat.mode & _0777).should.be.equal(_0755);
                }
                done();
            });
        });
    });
    it("should test relative path 3", function(done) {
        const x = Math.floor(Math.random() * Math.pow(16, 4)).toString(16);
        const y = Math.floor(Math.random() * Math.pow(16, 4)).toString(16);
        const z = Math.floor(Math.random() * Math.pow(16, 4)).toString(16);

        const file = nodePath.join(".", [x, y, z].join(nodePath.sep));

        mkdirp.mkdirp(file, _0755, function(err) {
            (err === null).should.be.equal(true);
            enFs.stat(file, function(errStat, stat) {
                (errStat === null).should.be.equal(true);
                stat.isDirectory().should.be.equal(true);
                if (!isWindows) {
                    (stat.mode & _0777).should.be.equal(_0755);
                }
                done();
            });
        });
    });
    it("should test return value", function(done) {
        const x = Math.floor(Math.random() * Math.pow(16, 4)).toString(16);
        const y = Math.floor(Math.random() * Math.pow(16, 4)).toString(16);
        const z = Math.floor(Math.random() * Math.pow(16, 4)).toString(16);

        const file = nodePath.join(tmpPath, [x, y, z].join(nodePath.sep));

        // should always return the full path created.
        //on second test it won't recreate the folder but will return the path
        mkdirp.mkdirp(file, function(err, made) {
            (err === null).should.be.equal(true);
            made.should.be.equal(nodePath.join(tmpPath, x, y, z));
            mkdirp.mkdirp(file, function(err, made2) {
                (err === null).should.be.equal(true);
                made2.should.be.equal(file);
                done();
            });
        });
    });
    it("should test root", function(done) {
        const file = nodePath.resolve(nodePath.sep);
        // '/' on unix, 'c:/' on windows.

        mkdirp.mkdirp(file, _0755, function(err) {
            (err === null).should.be.equal(true);
            enFs.stat(file, function(errStat, stat) {
                (err === null).should.be.equal(true);
                stat.isDirectory().should.be.equal(true);
                done();
            });
        });
    });
    it("should test implicit mode from umask", function(done) {
        const x = Math.floor(Math.random() * Math.pow(16, 4)).toString(16);
        const y = Math.floor(Math.random() * Math.pow(16, 4)).toString(16);
        const z = Math.floor(Math.random() * Math.pow(16, 4)).toString(16);

        const file = nodePath.join(tmpPath, [x, y, z].join(nodePath.sep));

        mkdirp.mkdirp(file, function(err) {
            (err === null).should.be.equal(true);
            enFs.stat(file, function(errStat, stat) {
                (err === null).should.be.equal(true);
                stat.isDirectory().should.be.equal(true);
                if (!isWindows) {
                    (stat.mode & _0777).should.be.equal(_0777);
                }
                done();
            });
        });
    });
    it("should test null byte filename", function(done) {
        const nullChar = "\0";

        const x = Math.floor(Math.random() * Math.pow(16, 4)).toString(16);
        const y = Math.floor(Math.random() * Math.pow(16, 4)).toString(16);
        const z = Math.floor(Math.random() * Math.pow(16, 4)).toString(16);

        const file = nodePath.join(tmpPath, [x, y, z].join(nodePath.sep), nullChar);

        mkdirp.mkdirp(file, function(err) {
            (err === null).should.be.equal(false);
            err.should.instanceof(Error);
            if (err.code) {
                err.code.should.be.equal("ENOENT");
            }
            err.message.should.containEql("string without null bytes");
            done();
        });
    });
    it("should test curly braces async", function(done) {
        const file = `${nodePath.resolve(tmpPath)}/{production,dev}/{css,img,js}`;
        const paths = mkdirpUtil.inspectCurlyBraces(file).map(function(path) {
            return nodePath.resolve(path);
        });
        let size = paths.length;
        mkdirp.mkdirp(file, function(err) {
            (err === null).should.be.equal(true);
            paths.forEach(function(path) {
                enFs.stat(path, function(errStat, stat) {
                    (errStat === null).should.be.equal(true);
                    stat.isDirectory().should.be.equal(true);
                    if (--size === 0) {
                        done();
                    }
                });
            });
        });
    });
    it("should test invalid path drive on windows", function(done) {
        const file = nodePath.join(invalidWindowsDrive, "fooAsync");
        if (!isWindows) {
            return done();
        }
        mkdirp.mkdirp(file, function(err) {
            if (isWindows) {
                (err === null).should.be.equal(false);
                err.code.should.be.equal("EINVALID");
                err.message.should.containEql("Invalid character found in path.");
            } else {
                (err === null).should.be.equal(true);
            }
            done();
        });
    });
    it("should test invalid filename with double quote async", function(done) {
        const file = nodePath.join(tmpPath, `foo"bar`);

        mkdirp.mkdirp(file, function(err) {
            if (isWindows) {
                (err === null).should.be.equal(false);
                //err.code.should.be.equal("ENOENT");
                err.message.should.containEql("Invalid character found in path");
            } else {
                (err === null).should.be.equal(true);
            }
            done();
        });
    });
    it("should test Date.toISOString", function(done) {
        const file = nodePath.join(tmpPath, (new Date()).toISOString(), "test");
        mkdirp.mkdirp(file, function(err) {
            if (!isWindows) {
                (err === null).should.be.equal(true);
                enFs.stat(file, function(errStat, stat) {
                    (errStat === null).should.be.equal(true);
                    stat.should.have.property("ctime");
                    //stat.should.be.instanceof(enFs.Stats);
                    stat.isDirectory().should.be.equal(true);
                    done();
                });
            } else {
                (err === null).should.be.equal(false);
                done();
            }
        });
    });
    it("should test mkdir with array of paths", function(done) {
        const file = [nodePath.join(tmpPath, "abc"), nodePath.join(tmpPath, "xyz")];

        let size = file.length;
        mkdirp.mkdirp(file, function(err) {
            (err === null).should.be.equal(true);
            file.forEach(function(path) {
                enFs.stat(path, function(errStat, stat) {
                    (errStat === null).should.be.equal(true);
                    stat.isDirectory().should.be.equal(true);
                    if (--size === 0) {
                        done();
                    }
                });
            });
        });
    });
    it("should test mkdir with array of paths and curly braces", function(done) {
        let paths = [];
        const x = Math.floor(Math.random() * Math.pow(16, 4)).toString(16);
        const y = Math.floor(Math.random() * Math.pow(16, 4)).toString(16);

        let file = [`${nodePath.join(tmpPath, x)}/{production,dev}/{css,img,js}`, `${nodePath.join(tmpPath, y)}/{production,dev}/{css,img,js}`];
        file.forEach(function(p) {
            paths = paths.concat(mkdirpUtil.inspectCurlyBraces(p));
        });
        let size = file.length;
        mkdirp.mkdirp(file, function(err) {
            (err === null).should.be.equal(true);
            paths.forEach(function(path) {
                enFs.stat(path, function(errStat, stat) {
                    (errStat === null).should.be.equal(true);
                    stat.isDirectory().should.be.equal(true);
                    if (--size === 0) {
                        done();
                    }
                });
            });
        });
    });
    it("should test mkdir with a file of the same name", function(done) {
        const x = Math.floor(Math.random() * Math.pow(16, 4)).toString(16);
        const file = nodePath.join(tmpPath, x, "file.json");
        enFs.mkdir(nodePath.dirname(file), function(errMkdir) {
            (errMkdir === null).should.be.equal(true);
            enFs.writeFile(file, "Data inside file", function(errWrite) {
                (errWrite === null).should.be.equal(true);
                mkdirp.mkdirp(file, function(err) {
                    err.should.be.instanceOf(Error);
                    err.code.should.be.equal("EEXIST");
                    done();
                });
            });
        });
    });
});


lib/util.js
/**
 * @project enfsmkdirp
 * @filename util.js
 * @description helper methods for enfsmkdirp module
 * @author Joao Parreira <joaofrparreira@gmail.com>
 * @copyright Copyright(c) 2016 Joao Parreira <joaofrparreira@gmail.com>
 * @licence Creative Commons Attribution 4.0 International License
 * @createdAt Created at 18-02-2016.
 * @version 0.0.2
 */

"use strict";

const nodePath = require("path");
const enFs = require("enfspatch");
const braceExpansion = require("brace-expansion");



const kindOf = (arg) => arg === null ? "null" : typeof arg === "undefined" ? "undefined" : /^\[object (.*)\]$/.exec(Object.prototype.toString.call(arg))[1].toLowerCase();
const isObject = (arg) => arg !== null && typeof arg !== "undefined" && "object" === kindOf(arg);
const isString = (arg) => "string" === kindOf(arg);
const inspectCurlyBraces = (path) => braceExpansion(path);



function getOptions(opt) {
    let options = {};

    if (opt) {
        if (!isObject(opt)) {
            options = {mode: opt};
        } else {
            options = opt;
        }
    }
    options.fs = options.fs || enFs;
    options.mode = options.mode ? isString(options.mode) ? parseInt(options.mode, 8) : options.mode : parseInt("0777", 8);

    return options;
}

function getPaths(originalPath) {
    let paths = [];

    const path = Array.isArray(originalPath) ? originalPath : [originalPath];

    path.forEach(function(p) {
        let ps = [p];
        if (p.indexOf("{")) {
            ps = inspectCurlyBraces(p);
        }
        ps = Array.isArray(ps) ? ps : [ps];
        paths = paths.concat(ps);
    });

    paths = paths.map(function(p) {
        return nodePath.resolve(p);
    });

    return paths;
}


module.exports = {
    inspectCurlyBraces,
    getOptions,
    getPaths
};



lib/mkdirpAsync.js
/**
 * @project enfsmkdirp
 * @filename mkdirpAsync.js
 * @description async methods for creating directories
 * @author Joao Parreira <joaofrparreira@gmail.com>
 * @copyright Copyright(c) 2016 Joao Parreira <joaofrparreira@gmail.com>
 * @licence Creative Commons Attribution 4.0 International License
 * @createdAt Created at 18-02-2016.
 * @version 0.0.2
 */

"use strict";


const nodePath = require("path");
const mkdirUtil = require("./util");
const isWindows = /^win/.test(process.platform);

function noop() {
}


function mkdirp_(path, options, callback) {
    options.fs.mkdir(path, options.mode, (err) => {
        if (err) {
            if (err.code !== "ENOENT" && err.code !== "EEXIST") {
                return options.fs.stat(path, (errStat, stat) => {
                    return callback((errStat || !stat.isDirectory()) ? err : null, errStat ? null : path);
                });
            } else if (err.code === "EEXIST") {
                return options.fs.stat(path, (errStat, stat) => {
                    return callback((errStat || !stat.isDirectory()) ? err : null, errStat ? null : path);
                });
            }
        } else {
            return callback(null, path);
        }

        //ENOENT
        const parent = nodePath.dirname(path);
        if (parent === path || /\0/.test(path)) {
            return callback(err);
        }
        if ((/"/.test(path) || /\:\\.*\\.*:.*\\/.test(path)) && isWindows) {
            let e = new Error("Invalid character found in path.");
            e.code = "EINVALID";
            return callback(e);
        }

        mkdirp_(parent, options, (errParent) => {
            if (errParent) {
                return callback(errParent);
            }
            options.fs.mkdir(path, options.mode, (errPath) => {
                //Ignore EEXIST error in asynchronous calls to mkdirp
                if (errPath && errPath.code !== "EEXIST") {
                    return callback(err);
                }
                callback(null, path);
            });
        });
    });
}

/**
 * Automatically create directories and sub-directories (mkdir -p)
 * @param {string|Array} path - the path to the directory
 * can be   . "/path/to/directory"
 *          . "/path/{to1,to2}/{dir1,dir2}"
 *          , ["/path/to/dir1","path/to/dir2"]
 * @param {object} opt - various options for list module
 *              {string|Number} options.mode - the mode to be affected to the directory
 *              {object} options.fs - the fs module to be used
 * @param {function} callback - the callback function that will be called after the list is done
 * @return {Error|string} Error | the last created path
 */
function mkdirp(path, opt, callback) {

    if (typeof opt === "function") {
        callback = opt;
        opt = {};
    }
    callback = callback || noop;

    const options = mkdirUtil.getOptions(opt);
    const paths = mkdirUtil.getPaths(path);

    const umask = process.umask();
    process.umask(0);
    const cwd = process.cwd();

    let size = paths.length;
    const done = [];
    let callbackCalled = false;
    paths.some((p) => {
        p = nodePath.resolve(cwd, p);
        if (callbackCalled) {
            return true;
        }
        mkdirp_(p, options, (err, m) => {
            if (err) {
                callbackCalled = true;
                process.umask(umask);
                return callback(err);
            }
            done.push(m);
            if (!--size && !callbackCalled) {
                callbackCalled = true;
                process.umask(umask);
                return callback(null, done.length > 1 ? done : done[0]);
            }
        });
    });
}


module.exports = mkdirp;
